www_src/pages/section-2/grid.jsx

```diff
@@ -0,0 +1,382 @@
+
+    var scaleTransform = 'scale(' + (this.displayState.zoom || 1) + ')';
+    var previousScaleTransform = 'scale(' + (this.displayState.previousZoom || 1) + ')';
+
+    var translateTransform = ' translate3d(' + this.displayState.cameraX + 'px, ' + this.displayState.cameraY + 'px, 0)';
+    var previousTranslateTransform = ' translate3d(' + this.displayState.previousCameraX + 'px, ' + this.displayState.previousCameraY + 'px, 0)';
+
+    var finalTransform = scaleTransform;
+
+    if (typeof this.displayState.cameraX === 'number' && typeof this.displayState.cameraY === 'number') {
+      finalTransform += (' '  + translateTransform)
+    }
+
+    if (animateZoom) {
+      elGrid.classList.add('animated');
+
+      elGrid.style.transform = previousScaleTransform + ' ' + translateTransform;
+
+      // TODO - use transition end event instead of timeout
+      setTimeout(function() {
+        elGrid.style.transform = finalTransform;
+
+        setTimeout(function() {
+          elGrid.classList.remove('animated');
+        }, 300);
+      }, 300);
+    } else {
+      elGrid.classList.remove('animated');
+      elGrid.style.transform = scaleTransform + ' ' + previousTranslateTransform;
+
+      // Allow previous transform to complete (prevent zoom throb motion at edges)
+      setTimeout(function() {
+        elGrid.classList.add('animated');
+        elGrid.style.transform = scaleTransform + ' ' + translateTransform;
+
+        setTimeout(function() {
+          elGrid.classList.remove('animated');
+        }, 300);
+      }, 1);
+    }
+  },
+  render: function () {
+    var self = this;
+
+    var nodes = [];
+    var layout = this.state.layout;
+
+    this.tilesPerRow = layout[0].length;
+    this.tilesPerCol = layout.length;
+
+    // Parse aspect ratio
+    var widthAR = parseInt(this.props.aspectRatio.split(':')[0], 10);
+    var heightAR = parseInt(this.props.aspectRatio.split(':')[1], 10);
+
+    // Try to fit grid in viewport by constraining to the width
+    this.slotWidth = (self.state.containerWidth / this.tilesPerRow);
+    this.slotHeight = this.slotWidth * (heightAR / widthAR);
+
+    // If the height overflows, then constrain by height instead
+    if (this.slotHeight * this.tilesPerCol > self.state.containerHeight) {
+      this.slotHeight = self.state.containerHeight / this.tilesPerCol;
+      this.slotWidth = this.slotHeight * (widthAR / heightAR);
+    }
+
+    this.slotWidth = Math.floor(this.slotWidth);
+    this.slotHeight = Math.floor(this.slotHeight);
+
+    var slotStyle = {
+      width: this.slotWidth + 'px',
+      height: this.slotHeight + 'px'
+    }
+
+    this.gridWidth = this.slotWidth * this.tilesPerRow;
+    this.gridHeight = this.slotHeight * this.tilesPerCol;
+
+    // Render slots
+    for (var y = 0; y < layout.length; y++) {
+      for (var x = 0; x < layout[0].length; x++) {
+        if (layout[y][x]) {
+          nodes.push(
+            <Slot x={x} y={y} style={slotStyle} key={ y + '-' + x }>
```

Comment:
all thee Slot generation blocks use mostly the same code, so there's a fair bit of cleanup we can do here, especially because JSX uses ES6 style property dereferencing:
```
var key = [y,x].join('-');
var slotProps = { x: x, y: y, style: slotStyle, key: key };
var slotContent = false;
var clickHandler = this.onPageClick.bind(this, {x:x, y:y});
if (layout[y][x]) {
  slotContent = <Page onClick={ clickHandler } screenshot={layout[y][x]} />
} else if (...) {
  slotContent = <AddPage onClick={ clickHandler } />
}
nodes.push(<Slot {...slotProps}>{ slotContent }</Slot>);
```